src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1851 @@
+/// given flags to DuplicateHandle.
+///
+/// Unlike win32 DuplicateHandle, this will preserve INVALID_HANDLE_VALUE (which is
+/// also the pseudohandle for the current process).
+fn dup_handle_to_process_with_flags(handle: &WinHandle, other_process: &WinHandle, flags: winapi::shared::minwindef::DWORD)
+                                           -> Result<WinHandle, WinError>
+{
+    if !handle.is_valid() {
+        return Ok(WinHandle::invalid());
+    }
+
+    unsafe {
+        let mut new_handle: HANDLE = INVALID_HANDLE_VALUE;
+        let ok = winapi::um::handleapi::DuplicateHandle(CURRENT_PROCESS_HANDLE.as_raw(), handle.as_raw(),
+                                           other_process.as_raw(), &mut new_handle,
+                                           0, winapi::shared::minwindef::FALSE, flags);
+        if ok == winapi::shared::minwindef::FALSE {
+            Err(WinError::last("DuplicateHandle"))
+        } else {
+            Ok(WinHandle::new(new_handle))
+        }
+    }
+}
+
+/// Duplicate a handle in the current process.
+fn dup_handle(handle: &WinHandle) -> Result<WinHandle,WinError> {
+    dup_handle_to_process(handle, &WinHandle::new(CURRENT_PROCESS_HANDLE.as_raw()))
+}
+
+/// Duplicate a handle to the target process.
+fn dup_handle_to_process(handle: &WinHandle, other_process: &WinHandle) -> Result<WinHandle,WinError> {
+    dup_handle_to_process_with_flags(handle, other_process, winapi::um::winnt::DUPLICATE_SAME_ACCESS)
+}
+
+/// Duplicate a handle to the target process, closing the source handle.
+fn move_handle_to_process(handle: WinHandle, other_process: &WinHandle) -> Result<WinHandle,WinError> {
+    let result = dup_handle_to_process_with_flags(&handle, other_process,
+                                                  winapi::um::winnt::DUPLICATE_CLOSE_SOURCE | winapi::um::winnt::DUPLICATE_SAME_ACCESS);
+    // Since the handle was moved to another process, the original is no longer valid;
+    // so we probably shouldn't try to close it explicitly?
+    mem::forget(handle);
+    result
+}
+
+#[derive(Debug)]
+struct WinHandle {
+    h: HANDLE
+}
+
+unsafe impl Send for WinHandle { }
+unsafe impl Sync for WinHandle { }
+
+impl Drop for WinHandle {
+    fn drop(&mut self) {
+        unsafe {
+            if self.is_valid() {
+                let result = winapi::um::handleapi::CloseHandle(self.h);
+                assert!(thread::panicking() || result != 0);
+            }
+        }
+    }
+}
+
+impl Default for WinHandle {
+    fn default() -> WinHandle {
+        WinHandle { h: INVALID_HANDLE_VALUE }
+    }
+}
+
+impl PartialEq for WinHandle {
+    fn eq(&self, other: &WinHandle) -> bool {
+        // WPre-Windows 10 API GetFileInformationByHandle() is used to
+        // to compare the underlying objects the handles refer to,
+        //
+        // On Windows 10, we could use:
+        // ```
+        // unsafe { winapi::um::handleapi::CompareObjectHandles(self.h, other.h) == winapi::shared::minwindef::TRUE }
```

Comment:
It looks like we don't actually use `CompareObjectHandles` (all uses are commented out), so this shouldn't lead to link errors (as reported in https://github.com/retep998/winapi-rs/issues/781). Any idea what may be causing that?
@jdm @angelortiz1007 @retep998

Reply:
I'm sorry for the noise, it's clearly defined in winapi-rs https://docs.rs/winapi/0.3.7/src/winapi/um/handleapi.rs.html#23-26, whereas I thought the snippets are for illustration purposes only. Will continue in the linked issue